lib/plugin/videoroom.ex

```diff
@@ -1,68 +1,282 @@
-    message = new_room_message(room_name, room_properties, session_id, handle_id)
-
-    case Connection.call(connection, message) do
-      {:ok, %{"videoroom" => "created", "room" => id}} ->
-        {:ok, id}
-
-      {:ok, %{"error_code" => 427, "videoroom" => "event"}} ->
-        {:error, :already_exists}
-
-      {:error, reason} ->
-        {:error, reason}
+  @moduledoc """
+  This module provides an API to manipulate and read properties of Janus Gateway's VideoRoom plugin's rooms.
+  """
+  alias Janus.Session
+  alias Janus.Plugin.VideoRoom.Errors
+  alias Janus.Plugin.VideoRoom.{CreateRoomProperties, EditRoomProperties}
+
+  @admin_key :admin_key
+  @secret_key :secret
+
+  @type room_id :: String.t()
+  @type room :: map
+  @type admin_key :: String.t() | nil
+  @type room_secret :: String.t() | nil
+  @type action :: String.t()
+  @type allowed :: list(String.t())
+
+  @type participant :: map
+
+  @doc """
+  Sends request to create a new room.
+
+  ## Arguments
+  * `session` - valid `Janus.Session` process to send request through
+  * `room_id` - new room's id, if set to nil gateway will create one
+  * `room_properties` - struct containing necessary information about room, see `t:Janus.Plugin.VideoRoom.CreateRoomProperties.t/0`
+  * `handle_id` - an id of caller's handle
+  * `admin_key` - optional admin key if gateway requires it
+  * `room_secret` - optional room secret when requested room has to be protected
+  """
+  @spec create_room(
+          Session.t(),
+          room_id,
+          CreateRoomProperties.t(),
+          Session.plugin_handle_id(),
+          admin_key,
+          room_secret
+        ) :: {:ok, String.t()} | {:error, any}
+  def create_room(
+        session,
+        room_id,
+        room_properties,
+        handle_id,
+        admin_key \\ nil,
+        room_secret \\ nil
+      ) do
+    message = configure(room_id, room_properties, handle_id, admin_key, room_secret, "create")
+
+    with {:ok, %{"videoroom" => "created", "room" => id}} <-
+           Session.execute_request(session, message) do
+      {:ok, id}
     end
+    |> Errors.handle_videoroom_error()
   end
 
-  defp new_room_message(room_id, room_properties, session_id, handle_id) do
+  @doc """
+  Sends request to edit given room.
+
+  ## Arguments
+  * `session` - valid `Janus.Session` process to send request through
+  * `room_id` - an id of targeted room
+  * `room_properties` - struct containing room information to be updated, see `t:Janus.Plugin.VideoRoom.EditRoomProperties.t/0`
+  * `handle_id` - an id of caller's handle
+  * `room_secret` - optional room secret when requested room is protected
+  """
+  @spec edit(Janus.Session.t(), room_id, EditRoomProperties.t(), Session.plugin_handle_id(), room_secret) ::
+          {:ok, room_id} | {:error, any}
+  def edit(
+        session,
+        room_id,
+        room_properties,
+        handle_id,
+        room_secret \\ nil
```

Comment:
I would opt for leaving those fields in `EditRoomProperties` and `CreateRoomProperties` just for the sake of knowing what `VideoRoom` might expect as request.

Reply:
In case of `CreateRoomProperties` - totally agree, it's part of the properties when setting up room. However, for `Edit...` it is not consistent - It does not have `room` entry, you're passing it separately, so your argument is invalid. For the exact shape of request - that is the purpose of Janus Docs. This also reminded me, that you should add links to Janus docs in modules with `Properties` structs